tensorflow/compiler/tf2xla/frontend_attributes_util.cc

```diff
@@ -0,0 +1,43 @@
+/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+#include "tensorflow/compiler/tf2xla/frontend_attributes_util.h"
+
+#include "tensorflow/core/framework/node_def_util.h"
+#include "tensorflow/core/lib/core/errors.h"
+
+namespace tensorflow {
+
+namespace {
+const char kFrontendAttributesAttribute[] = "_XlaFrontendAttributes";
+}  // namespace
+
+xla::StatusOr<absl::optional<xla::FrontendAttributes>>
+GetFrontendAttributesFromNodeDef(const NodeDef& node_def) {
+  if (!HasNodeAttr(node_def, kFrontendAttributesAttribute)) {
+    return absl::optional<xla::FrontendAttributes>();
+  }
+  string value;
+  xla::FrontendAttributes attributes;
+  TF_RETURN_IF_ERROR(
+      GetNodeAttr(node_def, kFrontendAttributesAttribute, &value));
+  if (!attributes.ParseFromString(value)) {
+    return errors::InvalidArgument(
+        "Experimental _XlaFrontendAttributes attribute was not a valid encoded "
+        "xla::FrontendAttributes proto.");
+  }
+  return absl::optional<xla::FrontendAttributes>(attributes);
```

Comment:
Can this be just `return attributes;`?

Reply:
No I don't think it can, probably for the same reason as above: the optional is embedded inside a StatusOr and as a result the conversion needs to be explicit:
```
ERROR: tensorflow/compiler/tf2xla/frontend_attributes_util.cc: In function 'xla::StatusOr<absl::optional<xla::FrontendAttributes> > tensorflow::GetFrontendAttributesFromNodeDef(const tensorflow::NodeDef&)':
ERROR: tensorflow/compiler/tf2xla/frontend_attributes_util.cc:40:10: error: could not convert 'attributes' from 'xla::FrontendAttributes' to 'xla::StatusOr<absl::optional<xla::FrontendAttributes> > {aka stream_executor::port::StatusOr<absl::optional<xla::FrontendAttributes> >}'
ERROR:    return attributes;
```